// @react-navigation/core/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  BaseNavigationContainer: true,
  createNavigatorFactory: true,
  NavigationHelpersContext: true,
  NavigationContext: true,
  NavigationRouteContext: true,
  CurrentRenderContext: true,
  useNavigationBuilder: true,
  useNavigation: true,
  useRoute: true,
  useFocusEffect: true,
  useIsFocused: true,
  useNavigationState: true,
  getStateFromPath: true,
  getPathFromState: true,
  getActionFromState: true,
  getFocusedRouteNameFromRoute: true
};
Object.defineProperty(exports, "BaseNavigationContainer", {
  enumerable: true,
  get: function get() {
    return _BaseNavigationContainer.default;
  }
});
Object.defineProperty(exports, "createNavigatorFactory", {
  enumerable: true,
  get: function get() {
    return _createNavigatorFactory.default;
  }
});
Object.defineProperty(exports, "NavigationHelpersContext", {
  enumerable: true,
  get: function get() {
    return _NavigationHelpersContext.default;
  }
});
Object.defineProperty(exports, "NavigationContext", {
  enumerable: true,
  get: function get() {
    return _NavigationContext.default;
  }
});
Object.defineProperty(exports, "NavigationRouteContext", {
  enumerable: true,
  get: function get() {
    return _NavigationRouteContext.default;
  }
});
Object.defineProperty(exports, "CurrentRenderContext", {
  enumerable: true,
  get: function get() {
    return _CurrentRenderContext.default;
  }
});
Object.defineProperty(exports, "useNavigationBuilder", {
  enumerable: true,
  get: function get() {
    return _useNavigationBuilder.default;
  }
});
Object.defineProperty(exports, "useNavigation", {
  enumerable: true,
  get: function get() {
    return _useNavigation.default;
  }
});
Object.defineProperty(exports, "useRoute", {
  enumerable: true,
  get: function get() {
    return _useRoute.default;
  }
});
Object.defineProperty(exports, "useFocusEffect", {
  enumerable: true,
  get: function get() {
    return _useFocusEffect.default;
  }
});
Object.defineProperty(exports, "useIsFocused", {
  enumerable: true,
  get: function get() {
    return _useIsFocused.default;
  }
});
Object.defineProperty(exports, "useNavigationState", {
  enumerable: true,
  get: function get() {
    return _useNavigationState.default;
  }
});
Object.defineProperty(exports, "getStateFromPath", {
  enumerable: true,
  get: function get() {
    return _getStateFromPath.default;
  }
});
Object.defineProperty(exports, "getPathFromState", {
  enumerable: true,
  get: function get() {
    return _getPathFromState.default;
  }
});
Object.defineProperty(exports, "getActionFromState", {
  enumerable: true,
  get: function get() {
    return _getActionFromState.default;
  }
});
Object.defineProperty(exports, "getFocusedRouteNameFromRoute", {
  enumerable: true,
  get: function get() {
    return _getFocusedRouteNameFromRoute.default;
  }
});

var _routers = require("@react-navigation/routers");

Object.keys(_routers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _routers[key];
    }
  });
});

var _BaseNavigationContainer = _interopRequireDefault(require("./BaseNavigationContainer"));

var _createNavigatorFactory = _interopRequireDefault(require("./createNavigatorFactory"));

var _NavigationHelpersContext = _interopRequireDefault(require("./NavigationHelpersContext"));

var _NavigationContext = _interopRequireDefault(require("./NavigationContext"));

var _NavigationRouteContext = _interopRequireDefault(require("./NavigationRouteContext"));

var _CurrentRenderContext = _interopRequireDefault(require("./CurrentRenderContext"));

var _useNavigationBuilder = _interopRequireDefault(require("./useNavigationBuilder"));

var _useNavigation = _interopRequireDefault(require("./useNavigation"));

var _useRoute = _interopRequireDefault(require("./useRoute"));

var _useFocusEffect = _interopRequireDefault(require("./useFocusEffect"));

var _useIsFocused = _interopRequireDefault(require("./useIsFocused"));

var _useNavigationState = _interopRequireDefault(require("./useNavigationState"));

var _getStateFromPath = _interopRequireDefault(require("./getStateFromPath"));

var _getPathFromState = _interopRequireDefault(require("./getPathFromState"));

var _getActionFromState = _interopRequireDefault(require("./getActionFromState"));

var _getFocusedRouteNameFromRoute = _interopRequireDefault(require("./getFocusedRouteNameFromRoute"));

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
