// react-native-paper/lib/commonjs/styles/colors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cyan100 = exports.cyan50 = exports.lightBlueA700 = exports.lightBlueA400 = exports.lightBlueA200 = exports.lightBlueA100 = exports.lightBlue900 = exports.lightBlue800 = exports.lightBlue700 = exports.lightBlue600 = exports.lightBlue500 = exports.lightBlue400 = exports.lightBlue300 = exports.lightBlue200 = exports.lightBlue100 = exports.lightBlue50 = exports.blueA700 = exports.blueA400 = exports.blueA200 = exports.blueA100 = exports.blue900 = exports.blue800 = exports.blue700 = exports.blue600 = exports.blue500 = exports.blue400 = exports.blue300 = exports.blue200 = exports.blue100 = exports.blue50 = exports.indigoA700 = exports.indigoA400 = exports.indigoA200 = exports.indigoA100 = exports.indigo900 = exports.indigo800 = exports.indigo700 = exports.indigo600 = exports.indigo500 = exports.indigo400 = exports.indigo300 = exports.indigo200 = exports.indigo100 = exports.indigo50 = exports.deepPurpleA700 = exports.deepPurpleA400 = exports.deepPurpleA200 = exports.deepPurpleA100 = exports.deepPurple900 = exports.deepPurple800 = exports.deepPurple700 = exports.deepPurple600 = exports.deepPurple500 = exports.deepPurple400 = exports.deepPurple300 = exports.deepPurple200 = exports.deepPurple100 = exports.deepPurple50 = exports.purpleA700 = exports.purpleA400 = exports.purpleA200 = exports.purpleA100 = exports.purple900 = exports.purple800 = exports.purple700 = exports.purple600 = exports.purple500 = exports.purple400 = exports.purple300 = exports.purple200 = exports.purple100 = exports.purple50 = exports.pinkA700 = exports.pinkA400 = exports.pinkA200 = exports.pinkA100 = exports.pink900 = exports.pink800 = exports.pink700 = exports.pink600 = exports.pink500 = exports.pink400 = exports.pink300 = exports.pink200 = exports.pink100 = exports.pink50 = exports.redA700 = exports.redA400 = exports.redA200 = exports.redA100 = exports.red900 = exports.red800 = exports.red700 = exports.red600 = exports.red500 = exports.red400 = exports.red300 = exports.red200 = exports.red100 = exports.red50 = void 0;
exports.orange300 = exports.orange200 = exports.orange100 = exports.orange50 = exports.amberA700 = exports.amberA400 = exports.amberA200 = exports.amberA100 = exports.amber900 = exports.amber800 = exports.amber700 = exports.amber600 = exports.amber500 = exports.amber400 = exports.amber300 = exports.amber200 = exports.amber100 = exports.amber50 = exports.yellowA700 = exports.yellowA400 = exports.yellowA200 = exports.yellowA100 = exports.yellow900 = exports.yellow800 = exports.yellow700 = exports.yellow600 = exports.yellow500 = exports.yellow400 = exports.yellow300 = exports.yellow200 = exports.yellow100 = exports.yellow50 = exports.limeA700 = exports.limeA400 = exports.limeA200 = exports.limeA100 = exports.lime900 = exports.lime800 = exports.lime700 = exports.lime600 = exports.lime500 = exports.lime400 = exports.lime300 = exports.lime200 = exports.lime100 = exports.lime50 = exports.lightGreenA700 = exports.lightGreenA400 = exports.lightGreenA200 = exports.lightGreenA100 = exports.lightGreen900 = exports.lightGreen800 = exports.lightGreen700 = exports.lightGreen600 = exports.lightGreen500 = exports.lightGreen400 = exports.lightGreen300 = exports.lightGreen200 = exports.lightGreen100 = exports.lightGreen50 = exports.greenA700 = exports.greenA400 = exports.greenA200 = exports.greenA100 = exports.green900 = exports.green800 = exports.green700 = exports.green600 = exports.green500 = exports.green400 = exports.green300 = exports.green200 = exports.green100 = exports.green50 = exports.tealA700 = exports.tealA400 = exports.tealA200 = exports.tealA100 = exports.teal900 = exports.teal800 = exports.teal700 = exports.teal600 = exports.teal500 = exports.teal400 = exports.teal300 = exports.teal200 = exports.teal100 = exports.teal50 = exports.cyanA700 = exports.cyanA400 = exports.cyanA200 = exports.cyanA100 = exports.cyan900 = exports.cyan800 = exports.cyan700 = exports.cyan600 = exports.cyan500 = exports.cyan400 = exports.cyan300 = exports.cyan200 = void 0;
exports.white = exports.black = exports.grey900 = exports.grey800 = exports.grey700 = exports.grey600 = exports.grey500 = exports.grey400 = exports.grey300 = exports.grey200 = exports.grey100 = exports.grey50 = exports.blueGrey900 = exports.blueGrey800 = exports.blueGrey700 = exports.blueGrey600 = exports.blueGrey500 = exports.blueGrey400 = exports.blueGrey300 = exports.blueGrey200 = exports.blueGrey100 = exports.blueGrey50 = exports.brown900 = exports.brown800 = exports.brown700 = exports.brown600 = exports.brown500 = exports.brown400 = exports.brown300 = exports.brown200 = exports.brown100 = exports.brown50 = exports.deepOrangeA700 = exports.deepOrangeA400 = exports.deepOrangeA200 = exports.deepOrangeA100 = exports.deepOrange900 = exports.deepOrange800 = exports.deepOrange700 = exports.deepOrange600 = exports.deepOrange500 = exports.deepOrange400 = exports.deepOrange300 = exports.deepOrange200 = exports.deepOrange100 = exports.deepOrange50 = exports.orangeA700 = exports.orangeA400 = exports.orangeA200 = exports.orangeA100 = exports.orange900 = exports.orange800 = exports.orange700 = exports.orange600 = exports.orange500 = exports.orange400 = void 0;
const red50 = '#ffebee';
exports.red50 = red50;
const red100 = '#ffcdd2';
exports.red100 = red100;
const red200 = '#ef9a9a';
exports.red200 = red200;
const red300 = '#e57373';
exports.red300 = red300;
const red400 = '#ef5350';
exports.red400 = red400;
const red500 = '#f44336';
exports.red500 = red500;
const red600 = '#e53935';
exports.red600 = red600;
const red700 = '#d32f2f';
exports.red700 = red700;
const red800 = '#c62828';
exports.red800 = red800;
const red900 = '#b71c1c';
exports.red900 = red900;
const redA100 = '#ff8a80';
exports.redA100 = redA100;
const redA200 = '#ff5252';
exports.redA200 = redA200;
const redA400 = '#ff1744';
exports.redA400 = redA400;
const redA700 = '#d50000';
exports.redA700 = redA700;
const pink50 = '#fce4ec';
exports.pink50 = pink50;
const pink100 = '#f8bbd0';
exports.pink100 = pink100;
const pink200 = '#f48fb1';
exports.pink200 = pink200;
const pink300 = '#f06292';
exports.pink300 = pink300;
const pink400 = '#ec407a';
exports.pink400 = pink400;
const pink500 = '#e91e63';
exports.pink500 = pink500;
const pink600 = '#d81b60';
exports.pink600 = pink600;
const pink700 = '#c2185b';
exports.pink700 = pink700;
const pink800 = '#ad1457';
exports.pink800 = pink800;
const pink900 = '#880e4f';
exports.pink900 = pink900;
const pinkA100 = '#ff80ab';
exports.pinkA100 = pinkA100;
const pinkA200 = '#ff4081';
exports.pinkA200 = pinkA200;
const pinkA400 = '#f50057';
exports.pinkA400 = pinkA400;
const pinkA700 = '#c51162';
exports.pinkA700 = pinkA700;
const purple50 = '#f3e5f5';
exports.purple50 = purple50;
const purple100 = '#e1bee7';
exports.purple100 = purple100;
const purple200 = '#ce93d8';
exports.purple200 = purple200;
const purple300 = '#ba68c8';
exports.purple300 = purple300;
const purple400 = '#ab47bc';
exports.purple400 = purple400;
const purple500 = '#9c27b0';
exports.purple500 = purple500;
const purple600 = '#8e24aa';
exports.purple600 = purple600;
const purple700 = '#7b1fa2';
exports.purple700 = purple700;
const purple800 = '#6a1b9a';
exports.purple800 = purple800;
const purple900 = '#4a148c';
exports.purple900 = purple900;
const purpleA100 = '#ea80fc';
exports.purpleA100 = purpleA100;
const purpleA200 = '#e040fb';
exports.purpleA200 = purpleA200;
const purpleA400 = '#d500f9';
exports.purpleA400 = purpleA400;
const purpleA700 = '#aa00ff';
exports.purpleA700 = purpleA700;
const deepPurple50 = '#ede7f6';
exports.deepPurple50 = deepPurple50;
const deepPurple100 = '#d1c4e9';
exports.deepPurple100 = deepPurple100;
const deepPurple200 = '#b39ddb';
exports.deepPurple200 = deepPurple200;
const deepPurple300 = '#9575cd';
exports.deepPurple300 = deepPurple300;
const deepPurple400 = '#7e57c2';
exports.deepPurple400 = deepPurple400;
const deepPurple500 = '#673ab7';
exports.deepPurple500 = deepPurple500;
const deepPurple600 = '#5e35b1';
exports.deepPurple600 = deepPurple600;
const deepPurple700 = '#512da8';
exports.deepPurple700 = deepPurple700;
const deepPurple800 = '#4527a0';
exports.deepPurple800 = deepPurple800;
const deepPurple900 = '#311b92';
exports.deepPurple900 = deepPurple900;
const deepPurpleA100 = '#b388ff';
exports.deepPurpleA100 = deepPurpleA100;
const deepPurpleA200 = '#7c4dff';
exports.deepPurpleA200 = deepPurpleA200;
const deepPurpleA400 = '#651fff';
exports.deepPurpleA400 = deepPurpleA400;
const deepPurpleA700 = '#6200ea';
exports.deepPurpleA700 = deepPurpleA700;
const indigo50 = '#e8eaf6';
exports.indigo50 = indigo50;
const indigo100 = '#c5cae9';
exports.indigo100 = indigo100;
const indigo200 = '#9fa8da';
exports.indigo200 = indigo200;
const indigo300 = '#7986cb';
exports.indigo300 = indigo300;
const indigo400 = '#5c6bc0';
exports.indigo400 = indigo400;
const indigo500 = '#3f51b5';
exports.indigo500 = indigo500;
const indigo600 = '#3949ab';
exports.indigo600 = indigo600;
const indigo700 = '#303f9f';
exports.indigo700 = indigo700;
const indigo800 = '#283593';
exports.indigo800 = indigo800;
const indigo900 = '#1a237e';
exports.indigo900 = indigo900;
const indigoA100 = '#8c9eff';
exports.indigoA100 = indigoA100;
const indigoA200 = '#536dfe';
exports.indigoA200 = indigoA200;
const indigoA400 = '#3d5afe';
exports.indigoA400 = indigoA400;
const indigoA700 = '#304ffe';
exports.indigoA700 = indigoA700;
const blue50 = '#e3f2fd';
exports.blue50 = blue50;
const blue100 = '#bbdefb';
exports.blue100 = blue100;
const blue200 = '#90caf9';
exports.blue200 = blue200;
const blue300 = '#64b5f6';
exports.blue300 = blue300;
const blue400 = '#42a5f5';
exports.blue400 = blue400;
const blue500 = '#2196f3';
exports.blue500 = blue500;
const blue600 = '#1e88e5';
exports.blue600 = blue600;
const blue700 = '#1976d2';
exports.blue700 = blue700;
const blue800 = '#1565c0';
exports.blue800 = blue800;
const blue900 = '#0d47a1';
exports.blue900 = blue900;
const blueA100 = '#82b1ff';
exports.blueA100 = blueA100;
const blueA200 = '#448aff';
exports.blueA200 = blueA200;
const blueA400 = '#2979ff';
exports.blueA400 = blueA400;
const blueA700 = '#2962ff';
exports.blueA700 = blueA700;
const lightBlue50 = '#e1f5fe';
exports.lightBlue50 = lightBlue50;
const lightBlue100 = '#b3e5fc';
exports.lightBlue100 = lightBlue100;
const lightBlue200 = '#81d4fa';
exports.lightBlue200 = lightBlue200;
const lightBlue300 = '#4fc3f7';
exports.lightBlue300 = lightBlue300;
const lightBlue400 = '#29b6f6';
exports.lightBlue400 = lightBlue400;
const lightBlue500 = '#03a9f4';
exports.lightBlue500 = lightBlue500;
const lightBlue600 = '#039be5';
exports.lightBlue600 = lightBlue600;
const lightBlue700 = '#0288d1';
exports.lightBlue700 = lightBlue700;
const lightBlue800 = '#0277bd';
exports.lightBlue800 = lightBlue800;
const lightBlue900 = '#01579b';
exports.lightBlue900 = lightBlue900;
const lightBlueA100 = '#80d8ff';
exports.lightBlueA100 = lightBlueA100;
const lightBlueA200 = '#40c4ff';
exports.lightBlueA200 = lightBlueA200;
const lightBlueA400 = '#00b0ff';
exports.lightBlueA400 = lightBlueA400;
const lightBlueA700 = '#0091ea';
exports.lightBlueA700 = lightBlueA700;
const cyan50 = '#e0f7fa';
exports.cyan50 = cyan50;
const cyan100 = '#b2ebf2';
exports.cyan100 = cyan100;
const cyan200 = '#80deea';
exports.cyan200 = cyan200;
const cyan300 = '#4dd0e1';
exports.cyan300 = cyan300;
const cyan400 = '#26c6da';
exports.cyan400 = cyan400;
const cyan500 = '#00bcd4';
exports.cyan500 = cyan500;
const cyan600 = '#00acc1';
exports.cyan600 = cyan600;
const cyan700 = '#0097a7';
exports.cyan700 = cyan700;
const cyan800 = '#00838f';
exports.cyan800 = cyan800;
const cyan900 = '#006064';
exports.cyan900 = cyan900;
const cyanA100 = '#84ffff';
exports.cyanA100 = cyanA100;
const cyanA200 = '#18ffff';
exports.cyanA200 = cyanA200;
const cyanA400 = '#00e5ff';
exports.cyanA400 = cyanA400;
const cyanA700 = '#00b8d4';
exports.cyanA700 = cyanA700;
const teal50 = '#e0f2f1';
exports.teal50 = teal50;
const teal100 = '#b2dfdb';
exports.teal100 = teal100;
const teal200 = '#80cbc4';
exports.teal200 = teal200;
const teal300 = '#4db6ac';
exports.teal300 = teal300;
const teal400 = '#26a69a';
exports.teal400 = teal400;
const teal500 = '#009688';
exports.teal500 = teal500;
const teal600 = '#00897b';
exports.teal600 = teal600;
const teal700 = '#00796b';
exports.teal700 = teal700;
const teal800 = '#00695c';
exports.teal800 = teal800;
const teal900 = '#004d40';
exports.teal900 = teal900;
const tealA100 = '#a7ffeb';
exports.tealA100 = tealA100;
const tealA200 = '#64ffda';
exports.tealA200 = tealA200;
const tealA400 = '#1de9b6';
exports.tealA400 = tealA400;
const tealA700 = '#00bfa5';
exports.tealA700 = tealA700;
const green50 = '#e8f5e9';
exports.green50 = green50;
const green100 = '#c8e6c9';
exports.green100 = green100;
const green200 = '#a5d6a7';
exports.green200 = green200;
const green300 = '#81c784';
exports.green300 = green300;
const green400 = '#66bb6a';
exports.green400 = green400;
const green500 = '#4caf50';
exports.green500 = green500;
const green600 = '#43a047';
exports.green600 = green600;
const green700 = '#388e3c';
exports.green700 = green700;
const green800 = '#2e7d32';
exports.green800 = green800;
const green900 = '#1b5e20';
exports.green900 = green900;
const greenA100 = '#b9f6ca';
exports.greenA100 = greenA100;
const greenA200 = '#69f0ae';
exports.greenA200 = greenA200;
const greenA400 = '#00e676';
exports.greenA400 = greenA400;
const greenA700 = '#00c853';
exports.greenA700 = greenA700;
const lightGreen50 = '#f1f8e9';
exports.lightGreen50 = lightGreen50;
const lightGreen100 = '#dcedc8';
exports.lightGreen100 = lightGreen100;
const lightGreen200 = '#c5e1a5';
exports.lightGreen200 = lightGreen200;
const lightGreen300 = '#aed581';
exports.lightGreen300 = lightGreen300;
const lightGreen400 = '#9ccc65';
exports.lightGreen400 = lightGreen400;
const lightGreen500 = '#8bc34a';
exports.lightGreen500 = lightGreen500;
const lightGreen600 = '#7cb342';
exports.lightGreen600 = lightGreen600;
const lightGreen700 = '#689f38';
exports.lightGreen700 = lightGreen700;
const lightGreen800 = '#558b2f';
exports.lightGreen800 = lightGreen800;
const lightGreen900 = '#33691e';
exports.lightGreen900 = lightGreen900;
const lightGreenA100 = '#ccff90';
exports.lightGreenA100 = lightGreenA100;
const lightGreenA200 = '#b2ff59';
exports.lightGreenA200 = lightGreenA200;
const lightGreenA400 = '#76ff03';
exports.lightGreenA400 = lightGreenA400;
const lightGreenA700 = '#64dd17';
exports.lightGreenA700 = lightGreenA700;
const lime50 = '#f9fbe7';
exports.lime50 = lime50;
const lime100 = '#f0f4c3';
exports.lime100 = lime100;
const lime200 = '#e6ee9c';
exports.lime200 = lime200;
const lime300 = '#dce775';
exports.lime300 = lime300;
const lime400 = '#d4e157';
exports.lime400 = lime400;
const lime500 = '#cddc39';
exports.lime500 = lime500;
const lime600 = '#c0ca33';
exports.lime600 = lime600;
const lime700 = '#afb42b';
exports.lime700 = lime700;
const lime800 = '#9e9d24';
exports.lime800 = lime800;
const lime900 = '#827717';
exports.lime900 = lime900;
const limeA100 = '#f4ff81';
exports.limeA100 = limeA100;
const limeA200 = '#eeff41';
exports.limeA200 = limeA200;
const limeA400 = '#c6ff00';
exports.limeA400 = limeA400;
const limeA700 = '#aeea00';
exports.limeA700 = limeA700;
const yellow50 = '#fffde7';
exports.yellow50 = yellow50;
const yellow100 = '#fff9c4';
exports.yellow100 = yellow100;
const yellow200 = '#fff59d';
exports.yellow200 = yellow200;
const yellow300 = '#fff176';
exports.yellow300 = yellow300;
const yellow400 = '#ffee58';
exports.yellow400 = yellow400;
const yellow500 = '#ffeb3b';
exports.yellow500 = yellow500;
const yellow600 = '#fdd835';
exports.yellow600 = yellow600;
const yellow700 = '#fbc02d';
exports.yellow700 = yellow700;
const yellow800 = '#f9a825';
exports.yellow800 = yellow800;
const yellow900 = '#f57f17';
exports.yellow900 = yellow900;
const yellowA100 = '#ffff8d';
exports.yellowA100 = yellowA100;
const yellowA200 = '#ffff00';
exports.yellowA200 = yellowA200;
const yellowA400 = '#ffea00';
exports.yellowA400 = yellowA400;
const yellowA700 = '#ffd600';
exports.yellowA700 = yellowA700;
const amber50 = '#fff8e1';
exports.amber50 = amber50;
const amber100 = '#ffecb3';
exports.amber100 = amber100;
const amber200 = '#ffe082';
exports.amber200 = amber200;
const amber300 = '#ffd54f';
exports.amber300 = amber300;
const amber400 = '#ffca28';
exports.amber400 = amber400;
const amber500 = '#ffc107';
exports.amber500 = amber500;
const amber600 = '#ffb300';
exports.amber600 = amber600;
const amber700 = '#ffa000';
exports.amber700 = amber700;
const amber800 = '#ff8f00';
exports.amber800 = amber800;
const amber900 = '#ff6f00';
exports.amber900 = amber900;
const amberA100 = '#ffe57f';
exports.amberA100 = amberA100;
const amberA200 = '#ffd740';
exports.amberA200 = amberA200;
const amberA400 = '#ffc400';
exports.amberA400 = amberA400;
const amberA700 = '#ffab00';
exports.amberA700 = amberA700;
const orange50 = '#fff3e0';
exports.orange50 = orange50;
const orange100 = '#ffe0b2';
exports.orange100 = orange100;
const orange200 = '#ffcc80';
exports.orange200 = orange200;
const orange300 = '#ffb74d';
exports.orange300 = orange300;
const orange400 = '#ffa726';
exports.orange400 = orange400;
const orange500 = '#ff9800';
exports.orange500 = orange500;
const orange600 = '#fb8c00';
exports.orange600 = orange600;
const orange700 = '#f57c00';
exports.orange700 = orange700;
const orange800 = '#ef6c00';
exports.orange800 = orange800;
const orange900 = '#e65100';
exports.orange900 = orange900;
const orangeA100 = '#ffd180';
exports.orangeA100 = orangeA100;
const orangeA200 = '#ffab40';
exports.orangeA200 = orangeA200;
const orangeA400 = '#ff9100';
exports.orangeA400 = orangeA400;
const orangeA700 = '#ff6d00';
exports.orangeA700 = orangeA700;
const deepOrange50 = '#fbe9e7';
exports.deepOrange50 = deepOrange50;
const deepOrange100 = '#ffccbc';
exports.deepOrange100 = deepOrange100;
const deepOrange200 = '#ffab91';
exports.deepOrange200 = deepOrange200;
const deepOrange300 = '#ff8a65';
exports.deepOrange300 = deepOrange300;
const deepOrange400 = '#ff7043';
exports.deepOrange400 = deepOrange400;
const deepOrange500 = '#ff5722';
exports.deepOrange500 = deepOrange500;
const deepOrange600 = '#f4511e';
exports.deepOrange600 = deepOrange600;
const deepOrange700 = '#e64a19';
exports.deepOrange700 = deepOrange700;
const deepOrange800 = '#d84315';
exports.deepOrange800 = deepOrange800;
const deepOrange900 = '#bf360c';
exports.deepOrange900 = deepOrange900;
const deepOrangeA100 = '#ff9e80';
exports.deepOrangeA100 = deepOrangeA100;
const deepOrangeA200 = '#ff6e40';
exports.deepOrangeA200 = deepOrangeA200;
const deepOrangeA400 = '#ff3d00';
exports.deepOrangeA400 = deepOrangeA400;
const deepOrangeA700 = '#dd2c00';
exports.deepOrangeA700 = deepOrangeA700;
const brown50 = '#efebe9';
exports.brown50 = brown50;
const brown100 = '#d7ccc8';
exports.brown100 = brown100;
const brown200 = '#bcaaa4';
exports.brown200 = brown200;
const brown300 = '#a1887f';
exports.brown300 = brown300;
const brown400 = '#8d6e63';
exports.brown400 = brown400;
const brown500 = '#795548';
exports.brown500 = brown500;
const brown600 = '#6d4c41';
exports.brown600 = brown600;
const brown700 = '#5d4037';
exports.brown700 = brown700;
const brown800 = '#4e342e';
exports.brown800 = brown800;
const brown900 = '#3e2723';
exports.brown900 = brown900;
const blueGrey50 = '#eceff1';
exports.blueGrey50 = blueGrey50;
const blueGrey100 = '#cfd8dc';
exports.blueGrey100 = blueGrey100;
const blueGrey200 = '#b0bec5';
exports.blueGrey200 = blueGrey200;
const blueGrey300 = '#90a4ae';
exports.blueGrey300 = blueGrey300;
const blueGrey400 = '#78909c';
exports.blueGrey400 = blueGrey400;
const blueGrey500 = '#607d8b';
exports.blueGrey500 = blueGrey500;
const blueGrey600 = '#546e7a';
exports.blueGrey600 = blueGrey600;
const blueGrey700 = '#455a64';
exports.blueGrey700 = blueGrey700;
const blueGrey800 = '#37474f';
exports.blueGrey800 = blueGrey800;
const blueGrey900 = '#263238';
exports.blueGrey900 = blueGrey900;
const grey50 = '#fafafa';
exports.grey50 = grey50;
const grey100 = '#f5f5f5';
exports.grey100 = grey100;
const grey200 = '#eeeeee';
exports.grey200 = grey200;
const grey300 = '#e0e0e0';
exports.grey300 = grey300;
const grey400 = '#bdbdbd';
exports.grey400 = grey400;
const grey500 = '#9e9e9e';
exports.grey500 = grey500;
const grey600 = '#757575';
exports.grey600 = grey600;
const grey700 = '#616161';
exports.grey700 = grey700;
const grey800 = '#424242';
exports.grey800 = grey800;
const grey900 = '#212121';
exports.grey900 = grey900;
const black = '#000000';
exports.black = black;
const white = '#ffffff';
exports.white = white;
